### PR TITLE
Add dedicated type for storing image hashes

### PR DESCRIPTION
**Pull request by GnomedDev**
This converts the hex representation that is used across the API into a much smaller (56 vs 17) type which improves type safety, saves space, and reduces heap allocations (only during a clone, but deserialization could be made work without an allocation if we pull in an ArrayString dependency).

### CODE REVIEW

File: src/model/guild/mod.rs [hunk 1]
```diff
@@ -626,7 +626,7 @@ impl Guild {
     pub async fn create(
         http: impl AsRef<Http>,
         name: &str,
-        icon: Option<&str>,
+        icon: ImageHash,
```
Comment (arqunis) [inline src/model/guild/mod.rs]: Why was this changed to a plain `ImageHash`? Guilds can be created without an icon, so this should stay as `Option`al.

```suggestion
        icon: Option<ImageHash>,
```

File: src/model/misc.rs [hunk 2]
```diff
@@ -53,6 +51,108 @@ impl_from_str! {
     ChannelId, ChannelIdParseError, parse_channel;
 }
 
+/// An image hash returned from the Discord API.
+///
+/// Note: This is parsed into a compact form when constructed, then turned back
+/// into the cannonical hex representation used by the API when needed.
```
Comment (arqunis) [inline src/model/misc.rs]: Speaking of my previous comment, it is not immediately clear how this type should be constructed by the user if it is going to be passed down as an argument to functions like `Guild::create`. We should mention the `FromStr` implementation here with an example.

File: src/model/misc.rs [hunk 3]
```diff
@@ -53,6 +51,108 @@ impl_from_str! {
...
+        for byte in self.hash {
+            write!(f, "{byte:02x}")?;
+        }
+
+        Ok(())
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum ImageHashParseError {
```
Comment (arqunis) [inline src/model/misc.rs:140]: If `FromStr` will be the go-to method for constructing an `ImageHash`, the error type should be documented as well.

File: src/model/misc.rs [hunk 4]
```diff
@@ -53,6 +51,108 @@ impl_from_str! {
...
+}
+
+impl std::fmt::Display for ImageHashParseError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::InvalidLength(length) => {
+                write!(f, "Invalid length {length}, expected 32 or 34 characters")
+            },
+            Self::MissingAnimatedMark => f.write_str("Input is 34 characters long, but missing a_"),
+            Self::UnparsableBytes(bytes) => write!(f, "Could not parse to hex: {bytes}"),
```
Comment (arqunis) [inline src/model/misc.rs]: Types that implement `std::error::Error` should not be embedded within `Display`'d messages. Instead, they should be returned as [the `source` of the error](https://doc.rust-lang.org/nightly/std/error/trait.Error.html#method.source), so that error crates [like `color-eyre`](https://github.com/yaahc/color-eyre) can integrate with these types nicely. While we're on the topic, `ImageHashParseError` should implement `std::error::Error`.
```suggestion
            Self::UnparsableBytes(_) => f.write_str("Could not parse to hex"),
```

File: src/model/misc.rs [hunk 5]
```diff
@@ -53,6 +51,108 @@ impl_from_str! {
...
+impl<'de> serde::Deserialize<'de> for ImageHash {
+    fn deserialize<D: serde::Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
+        // TODO: Replace this with ArrayString<34>?
+        let helper = String::deserialize(deserializer)?;
+        Self::from_str(&helper).map_err(serde::de::Error::custom)
+    }
+}
+
+impl std::fmt::Display for ImageHash {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
```
Comment (arqunis) [inline src/model/misc.rs:121]: Additionally, if `Display` will the the go-to way for 'peeking' into an `ImageHash`, it should be documented how the output should look like.

File: src/model/misc.rs [hunk 6]
```diff
@@ -53,6 +51,108 @@ impl_from_str! {
...
+        let (hex, is_animated) = if s.len() == 34 {
+            if &s.as_bytes()[0..2] != b"a_" {
+                return Err(ImageHashParseError::MissingAnimatedMark);
+            }
+
+            (&s[2..], true)
+        } else if s.len() == 32 {
+            (s, false)
+        } else {
+            return Err(Self::Err::InvalidLength(s.len()));
```
Comment (arqunis) [inline src/model/misc.rs:179]: Pedantic, but for consistency.
```suggestion
            return Err(ImageHashParseError::InvalidLength(s.len()));
```

**Review by arqunis — CHANGES_REQUESTED** [review]

File: src/model/guild/mod.rs (hunk 1 above)
Comment (GnomedDev) [reply src/model/guild/mod.rs]: Ah, guess this was a mistake

**Review by arqunis — APPROVED** [review]
